### PR TITLE
feat(libnvme-rs): new crate, implement connect and disconnect

### PR DESCRIPTION
**Pull request by jonathan-teh**
New crate that uses the libnvme development C library. Reuse NvmeTarget, and some NvmeErrors, from nvmeadm and implement connect and disconnect to an NVMe-oF target using libnvme.

Fixes CAS-1218

### CODE REVIEW

**Comment by jonathan-teh** [comment]
bors try

**Comment by jonathan-teh** [comment]
Compared to #1047 this uses `libnvme` to find the NVMe namespace that it connected to and uses that to return the path of the block device. There is now an `NvmeTransportType` enum so there is no need to check for a valid `trtype` in `NvmeTarget::connect()`.

**Comment by jonathan-teh** [comment]
bors try

**Comment by jonathan-teh** [comment]
bors try

**Comment by jonathan-teh** [comment]
bors try

**Review by paulyoong — APPROVED** [review]

File: libnvme-rs/src/nvme_uri.rs [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        let ret = unsafe {
+            crate::nvmf_add_ctrl(h, nvme_ctrl, &cfg, cfg.disable_sqflow)
+        };
+        if ret != 0 {
+            return Err(NvmeError::AddCtrlrError {
+                rc: ret,
+            });
+        }
+
+        let mut retries = 10;
```
Comment (gila) [inline libnvme-rs/src/nvme_uri.rs:185]: What is the reason for the loop? We have the controller connect here, and now we try to find the namespaces? A controller can have zero namespaces.

File: libnvme-rs/src/nvme_uri.rs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+            nr_poll_queues: 0,
+            tos: -1,
+            duplicate_connect: false,
+            disable_sqflow: false,
+            hdr_digest: false,
+            data_digest: false,
+        };
+        let ret = unsafe {
+            crate::nvmf_add_ctrl(h, nvme_ctrl, &cfg, cfg.disable_sqflow)
+        };
```
Comment (gila) [inline libnvme-rs/src/nvme_uri.rs:178]: What is the contract here? ret == 0 means we have, connected to the controller -- and its namespaces? or just the controller?

File: libnvme-rs/src/nvme_uri.rs [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+                    std::ffi::CStr::from_ptr(crate::nvme_subsystem_get_nqn(
+                        subsys,
+                    ))
+                };
+                if cstr.to_str().unwrap() != self.subsysnqn {
+                    continue;
+                }
+                let ctrlriter = NvmeCtrlrIterator::new(subsys);
+                for c in ctrlriter {
+                    if unsafe { crate::nvme_disconnect_ctrl(c) == 0 } {
```
Comment (gila) [inline libnvme-rs/src/nvme_uri.rs:248]: What does != 0 mean here? cant be found? or, can be found but can't disconnect? or.... ? if the disconnect call would yield EAGAIN or EBUSY, then yeah I can imagine we MAY want to retry but perhaps it's better to handle that further up, and let the consumer decide -- not the library.

File: libnvme-rs/src/nvme_tree.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    root: *mut crate::bindings::nvme_root,
+}
+
+impl NvmeRoot {
+    pub fn new(root: *mut crate::bindings::nvme_root) -> Self {
+        NvmeRoot {
+            root,
+        }
+    }
+    pub fn as_mut_ptr(&self) -> *mut crate::bindings::nvme_root {
```
Comment (gila) [inline libnvme-rs/src/nvme_tree.rs:12]: The goal of the lib should be to NOT have to deal with these raw dangling pointers, why is this method here and public?

File: libnvme-rs/src/nvme_uri.rs (hunk 1 above)
Comment (jonathan-teh) [reply libnvme-rs/src/nvme_uri.rs:185]: That was carried over from `nvmeadm`, I guess it's to allow for slow connections to the target. The assumption here is that the target will have at least one namespace.

File: libnvme-rs/src/nvme_uri.rs (hunk 2 above)
Comment (jonathan-teh) [reply libnvme-rs/src/nvme_uri.rs:178]: This means it has successfully written to `/dev/nvme-fabrics` and read a response back, which AIUI just means it has connected to the controller.

File: libnvme-rs/src/nvme_uri.rs (hunk 3 above)
Comment (jonathan-teh) [reply libnvme-rs/src/nvme_uri.rs:248]: This basically writes `1` to `/sys/class/nvme/{}/delete_controller` so `!=0` would an IO error of some description. I'll try to pass the error back up.

File: libnvme-rs/src/nvme_tree.rs (hunk 4 above)
Comment (jonathan-teh) [reply libnvme-rs/src/nvme_tree.rs:12]: That's an oversight, it's meant to be `pub(crate)`, will fix.

File: libnvme-rs/src/nvme_uri.rs (hunk 2 above)
Comment (gila) [reply libnvme-rs/src/nvme_uri.rs:178]: I just looked into the function and it does a whole lot more than that. Perhaps we are looking at different versions but master seems to add the controller (or returns <0) and then goes on and configure it and adds the controller to the subsystem. From there,  it seems to me we can use nvme_ctrL-first_ns() would be a lot better than this manual looping.

Comment (jonathan-teh) [reply libnvme-rs/src/nvme_uri.rs:178]: `NvmeNamespaceIterator::next()` essentially calls `nvme_subsystem_first_ns()` or `nvme_subsystem_next_ns()`.

Comment (gila) [reply libnvme-rs/src/nvme_uri.rs:178]: Then I dont understand why the loop within intervals of 1 second, creates a new root and iterates over it. Once the controller is connected it is part of the root tree, and so we can just list the namespaces once, right?

Comment (jonathan-teh) [reply libnvme-rs/src/nvme_uri.rs:178]: AIUI the kernel connects to the namespace asynchronously after connecting to the controller. We have a similar race condition in one of the unit tests which uses nvme-cli where a `nvme list` immediately after `nvme connect` returns no NVMe devices.

File: libnvme-rs/src/nvme_uri.rs (hunk 3 above)
Comment (jonathan-teh) [reply libnvme-rs/src/nvme_uri.rs]: I've passed the error back up as just the raw return code as there's no easy way to tell where it failed.

**Pull request closed without merging**